hal/src/argon/platform_ncp_argon.cpp

```diff
@@ -56,36 +68,86 @@ class OtaUpdateSourceStream : public particle::InputStream {
         if (!(flags & InputStream::READABLE)) {
             return SYSTEM_ERROR_INVALID_ARGUMENT;
         }
-        if (!remaining) {
+        if (!remaining_) {
             return SYSTEM_ERROR_END_OF_STREAM;
         }
         return InputStream::READABLE;
     }
 
 private:
-    const uint8_t* buffer;
-    size_t remaining;
+    uintptr_t address_;
+    size_t offset_;
+    size_t remaining_;
+    ReadStreamFunc readFunc_;
 };
 
+int invalidateWifiNcpVersionCache() {
+    using namespace particle::services;
+    // Cache will be updated by the NCP client itself
+    LOG(TRACE, "Invalidating cached ESP32 NCP firmware version");
+    return SystemCache::instance().del(SystemCacheKey::WIFI_NCP_FIRMWARE_VERSION);
+}
+
+int getWifiNcpFirmwareVersion(uint16_t* ncpVersion) {
+    uint16_t version = 0;
+
+    using namespace particle::services;
+    int res = SystemCache::instance().get(SystemCacheKey::WIFI_NCP_FIRMWARE_VERSION, &version, sizeof(version));
+    if (res == sizeof(version)) {
+        LOG(TRACE, "Cached ESP32 NCP firmware version: %d", (int)version);
+        *ncpVersion = version;
+        return 0;
+    }
+
+    if (res >= 0) {
+        invalidateWifiNcpVersionCache();
+    }
+
+    // Not present in cache or caching not supported, call into NCP client
+    const auto ncpClient = particle::wifiNetworkManager()->ncpClient();
+    SPARK_ASSERT(ncpClient);
+    const particle::NcpClientLock lock(ncpClient);
+    const particle::NcpPowerState ncpPwrState = ncpClient->ncpPowerState();
+    CHECK(ncpClient->on());
+    CHECK(ncpClient->getFirmwareModuleVersion(&version));
+    *ncpVersion = version;
+    if (ncpPwrState == particle::NcpPowerState::OFF) {
+        CHECK(ncpClient->off());
+    }
+
+    return 0;
+}
+
+} // anonymous
+
 // FIXME: This function accesses the module info via XIP and may fail to parse it correctly under
 // some not entirely clear circumstances. Disabling compiler optimizations helps to work around
 // the problem
 __attribute__((optimize("O0"))) int platform_ncp_update_module(const hal_module_t* module) {
```

Comment:
This seams wrong. We have removed this in the XIP removal branch. Besides, most of the implementation had been moved to `hal/network/ncp/wifi/platform_ncp_update.cpp`

Reply:
Thanks for clarifying, @XuGuohui. I meant to post this question - Modified. We are not including `hal/network/ncp/wifi/platform_ncp_update.cpp` in 2.x line - please correct me if I am wrong.

Reply:
@keeramis Yes, we can keep things in argon-specific source file in 2.x as 2.x does not support ESP32 on Tracker (which is the reason why this functionality was moved into a more generic location).

Reply:
Any reason why we are not going to include the refactored `platform_ncp_update.cpp`?

Reply:
We are including the contents, but into `platform_ncp_argon.cpp`. No strong preferences honestly, but doesn't make sense to move things around, again because Tracker does not support WiFi in 2.x.

Reply:
That makes sense 👍